backend/src/main/java/fk/prof/backend/aggregator/AggregationWindow.java

```diff
@@ -69,27 +70,37 @@ public AggregationState abandonProfile(long workId) throws AggregationFailure {
   }
 
   /**
-   * Aborts all in-flight profiles. Should be called when aggregation window expires
+   * Called by backend daemon thread
+   * Does following tasks required to expire aggregation window:
+   * > Marks status of ongoing profiles as aborted
+   * > De associates assigned work with this aggregation window
+   * > Updates ended at time for aggregation window
+   * > Finalizes the window
+   * @param aggregationWindowLookupStore
+   * @return finalized aggregation window
    */
-  public void abortOngoingProfiles() throws AggregationFailure {
+  public FinalizedAggregationWindow expireWindow(AggregationWindowLookupStore aggregationWindowLookupStore) {
     ensureEntityIsWriteable();
 
-    for (Map.Entry<Long, ProfileWorkInfo> entry : workInfoLookup.entrySet()) {
-      try {
-        entry.getValue().abortProfile();
-      } catch (IllegalStateException ex) {
-        throw new AggregationFailure(String.format("Error aborting profile for work_id=%d", entry.getKey()), ex);
-      }
-    }
+    abortOngoingProfiles();
+    long[] workIds = this.workInfoLookup.keySet().stream().mapToLong(Long::longValue).toArray();
+    aggregationWindowLookupStore.deAssociateAggregationWindow(workIds);
+    this.endedAt = LocalDateTime.now(Clock.systemUTC());
```

Comment:
vertx delegates timer stuff to schedule methods exposed by netty's event loop. Netty converts this into a [ScheduledFutureTask](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/ScheduledFutureTask.java) and scheduling queue is handled by [AbstractScheduledEventExecutor](https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/concurrent/AbstractScheduledEventExecutor.java)
Netty uses system nanotime internally (stable clock), so wall clock skew on backend should not affect us. Of course, end time can still be recorded incorrectly, so additionally will store duration in the object as a separate field